Improvement: Add gologr svclog wrapper

etherandrius opened the pull request:
## After this PR
svc1log wrapper for the go-logr logger.

This is needed to get proper logs when using k8s sigs controller-runtime. https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/log/log.go#L42-L46
==COMMIT_MSG==
==COMMIT_MSG==

## Possible downsides?
<!-- Please describe any way users could be negatively affected by this PR. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114)
<!-- Reviewable:end -->

Commit:
Improvement: Add gologr svclog wrapper

Commit:
bump warnings to errors

etherandrius commented on gologr-wlog/gologr.go:
```diff
@@ -0,0 +1,69 @@
+
+func toSafeParams(logger svc1log.Logger, keysAndValues []interface{}) svc1log.Param {
+	if len(keysAndValues)%2 == 1 {
+		logger.Error("KeysAndValues pair slice has an odd number of arguments; ignoring all",
```
I've read this https://dave.cheney.net/2015/11/05/lets-talk-about-logging and decided to bump these up from `Warn` -> `Error`

splucs commented on gologr-wlog/gologr.go:
```diff
@@ -0,0 +1,69 @@
+package gologrwlog
```
We should follow the same pattern for naming interface converters for logr and just name the package `wlogr`.
Generally open source loggers that want to implement the logr interface simply name their implementation by adding an "r" at the end.
Examples:
- [klog](https://github.com/kubernetes/klog) implements [klogr](https://github.com/kubernetes/klog/tree/master/klogr).
- [zap](https://github.com/uber-go/zap) implements [zapr](https://github.com/go-logr/zapr).

splucs replied:
Also, this is not an converter for the [wlog.Logger](https://github.com/palantir/witchcraft-go-logging/blob/5c3826853c72e1b8f277ed09ad1ee655fa114951/wlog/logentry.go#L37) but rather for the [svc1log.Logger](https://github.com/palantir/witchcraft-go-logging/blob/5c3826853c72e1b8f277ed09ad1ee655fa114951/wlog/svclog/svc1log/logger.go#L23) interface. I believe this should live under the path `github.com/palantir/witchcraft-go-logging/wlog/svclog/svc1log/wlogr`.

splucs commented on gologr-wlog/gologr.go:
```diff
@@ -0,0 +1,69 @@
+}
+
+// NewGoLogrWrapper returns a go-logr Logger implementation that uses svc1log to write logs.
+func NewGoLogrWrapper(logger svc1log.Logger, origin string) logr.Logger {
```
Package name already tells what it is, just name this `New`.

splucs commented on gologr-wlog/gologr.go:
```diff
@@ -0,0 +1,69 @@
+		return svc1log.SafeParams(map[string]interface{}{})
+	}
+
+	params := map[string]interface{}{}
```
`make(map[string]interface{}, len(keysAndValues)/2)`

splucs commented on gologr-wlog/gologr_test.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+func assertLogLine(t *testing.T, logLine []byte, expectedLogEntry logEntry) {
+	fmt.Println(string(logLine))
```
Remove debug statements.

splucs commented on gologr-wlog/gologr_test.go:
```diff
@@ -0,0 +1,77 @@
+	_ "github.com/palantir/witchcraft-go-logging/wlog-zap"
+)
+
+type logEntry struct {
```
This is fragile as the API might break in case the `wlog-zap` logEntry implementation changes.
Can you implement a mock logger that sends logs to a channel and assert on these messages?

etherandrius replied:
Instead of using a struct I am now using a JSONMatcher

following the internal tests: https://github.com/etherandrius/witchcraft-go-logging/blob/develop/wlog/svclog/svc1log/svc1logtests/tests.go#L106-L125

In terms of fragility it's the same, but I think it's acceptable

splucs commented on gologr-wlog/gologr.go:
```diff
@@ -0,0 +1,69 @@
+	"github.com/palantir/witchcraft-go-logging/wlog/svclog/svc1log"
+)
+
+type goLogrWrapper struct {
```
Just name this `logger`.

splucs commented on gologr-wlog/gologr.go:
```diff
@@ -0,0 +1,69 @@
+	}
+}
+
+func (s *goLogrWrapper) Info(msg string, keysAndValues ...interface{}) {
```
Name the receiver the initial letter of the struct.

Commit:
Address CR

Commit:
more consistent naming

Commit:
tests

etherandrius commented:
Verify succeeded on the latest commit, but not showing up below.

Commit:
changelog

bmoylan reviewed (changes requested):
> Reviewed 10 of 12 files at r2.
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114)* status: 9 of 11 files reviewed, 9 unresolved discussions (waiting on @etherandrius and @splucs)

___
*[wlog/svclog/svc1log/svc1logr/svc1logr.go, line 15 at r2](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MRVivCP4wqreVg4KZHD:-MRVivCP4wqreVg4KZHE:b-lgu0ra) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/3ae166decedd679dcd1c557ed5a00b3d06acfa6d/wlog/svclog/svc1log/svc1logr/svc1logr.go#L15)):*
> ```Go
> // limitations under the License.
> 
> package svc1logr
> ```

So I'd actually argue that because this is ancillary to the core functionality and products import it directly, it should be a top-level package if it's going to be in this repository. 

`svc1logr` feels a little too succinct to indicate that this implements a different api and use case than the `svc1log` library. Don't want to spend a bunch of energy bikeshedding but something like `wlog-logr` would make more sense to me.

We should also add some thorough package-level documentation that describes what this is and why one might need to use it given that it will be relatively visible in this repo but only used in a handful of specific cases where we want compatibility. 


<!-- Sent from Reviewable.io -->

nmiyake commented:
Similar to @bmoylan's feedback, can you provide more context around how this will be used? I see the initial comment around "k8s sigs controller-runtime", but was hoping to get more context.

Is the idea that the top-level `main` package that uses those packages will import this logger implementation so that all logging goes through the same logging mechanism? Want to understand if this is a straightforward case of adding another logger implementation (like `zapper` or `zerolog` etc.) or if this is attempting to do something different.

etherandrius commented:
> Similar to @bmoylan's feedback, can you provide more context around how this will be used? I see the initial comment around "k8s sigs controller-runtime", but was hoping to get more context.

In general this should be used when you are using a library that uses the gologr log interface and you want to get logs from the library. For example we will use this to get logs from a conversion webhook implementation inside k8s sigs [links](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/webhook/conversion/conversion.go).

> Is the idea that the top-level main package that uses those packages will import this logger implementation so that all logging goes through the same logging mechanism?

Yes.

> Want to understand if this is a straightforward case of adding another logger implementation (like zapper or zerolog etc.) or if this is attempting to do something different.

The main difference is that the returned logger is not a wlog logger but gologr logger. So in a sense it's the opposite of what the other packages do. 

> svc1logr feels a little too succinct to indicate that this implements a different api and use case than the svc1log library. Don't want to spend a bunch of energy bikeshedding but something like wlog-logr would make more sense to me.

I like making it a top level package. However, I would prefer to name it `svc1logr` (or `logr-svc1log`). 
- I don't want the name to have `wlog` in it since the implementation is specific to `svc1log` (not evt2log, etc..).
- I don't want the name to be of the form `wlog-*` since the package is "opposite" of the other packages with this naming convention.

Commit:
move package to top level

generate

update readme

rename package

add package readme

rename receiver

svc1logr

nit

nmiyake reviewed (commented):
Discussed further offline and synthesizing summary here.

My main question/hesitation is around whether it makes sense to have adapters (functions that wrap an svc1log in another logger interface for the purpose of interoperability with other libraries) as part of this main project. Unlike the logger implementation providers, this use case seemed more specialized.

However, it does make sense that the service logger does map most closely to a "regular" logger, and in scenarios like this where an external library/package expects a particular logger interface and we want the ability to delegate the logging to service logs, centralizing the translation here could make sense.

Currently I'm leaning towards being okay with having this implementation here, but would like to change the layout/organization to make it more general and make it more straightforward in a case where we might want to add another logger adapter of this form. For example, @etherandrius flagged that a similar pattern could be used to create an adapter for the go-acme logger (https://github.com/go-acme/lego/blob/master/log/logger.go) for projects that expect that logger but where we want to log to a service logger.

I'm going to examine a few other repositories that perform this kind of conversion and will then respond with concrete feedback in terms of where to move code and implementation.

> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-:-MS9I6p45w4Eh-I2gUwB:b-bcmkvr)* status: 9 of 13 files reviewed, 9 unresolved discussions (waiting on @etherandrius and @splucs)



<!-- Sent from Reviewable.io -->

nmiyake reviewed (commented):
Ah interesting -- taking a look at https://github.com/go-logr, the org itself has a repository per implementation for things like "zapr", "stdr", etc.

Ideally I think the best approach would be to have a `wlogr` or `svc1logr` implementation repo there. However, I think the probability of getting that done is probably low (even though this repo is OSS, it's not widely used so not sure that the maintainer would accept a repository there, and presumably we would also have less control).

Based on this, wondering if it might make more sense to have a `svc1logr` module in `palantir/pkg` rather than including it here -- that would fit better with my intuition that this adapter is a bit more of a separate client-type thing than the logger implementations themselves, which we do package with this repository (although I suppose an argument could be made that even those may make sense as separate modules).

@bmoylan do you have any thoughts on the above? Concretely, what do you think between having something like `svclog/svc1log/adpaters` with `svc1logr` as one of the directories, vs. having a separate `github.com/palantir/pkg/svc1logr` module?

> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-:-MS9XQDU3XbpKKmJQREX:b-uru1yi)* status: 9 of 13 files reviewed, 9 unresolved discussions (waiting on @etherandrius and @splucs)



<!-- Sent from Reviewable.io -->

nmiyake reviewed (changes requested):
Provided review on the content. Besides the content, the high-level question is the one posed above as to whether it makes sense to proceed with this in this repository or as a separate module.

> Reviewed 10 of 12 files at r2, 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-:-MSA1YrQ3jh05BXdHHcz:b-4t4i7w)* status: all files reviewed, 14 unresolved discussions (waiting on @etherandrius and @splucs)

___
*[README.md, line 13 at r3](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MS9j_Am5vfL0UVeoZFV:-MS9j_An5bJtNzOE573l:b-jcxepy) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/2b396e73357953070924915e7a12a5d8a0e80100/README.md#L13)):*
> ```Markdown
> libraries. `witchcraft-go-logging` includes implementations that use [zap](https://github.com/uber-go/zap), 
> [zerolog](https://github.com/rs/zerolog) and [glog](https://github.com/golang/glog). We also provide an implementation
> for [go-logr](https://github.com/go-logr/logr) that uses svc1log internally.
> ```

If we proceed with this PR in this repository, then this should be expanded a bit -- I think it would be worth explaining that the repository also provides adapters that will return a particular logger implementation using svc1log underneath, and that an implementation for logr is provided.
___
*[gologr-wlog/gologr.go, line 52 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MQvZ_Sp27m0ErnCysEF-r1-52:-MSA0O6--kEli8KgmqXz:b-1fqkhx) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/c6be7af3baf0bbcd2b983273fdde7eeff1d0e0b7/gologr-wlog/gologr.go#L52)):*
<details><summary><i>Previously, etherandrius (Andrius Grabauskas) wrote…</i></summary><blockquote>

I've read this https://dave.cheney.net/2015/11/05/lets-talk-about-logging and decided to bump these up from `Warn` -> `Error`
</blockquote></details>

When this condition is hit, is it clear that the error is in the logging layer? Agree that it's good to message this in some way, but want to be sure that the source of the error is clear.
___
*[svc1logr/svc1logr.go, line 31 at r3](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MSA19kR6HuWqGXUEaej:-MSA19kR6HuWqGXUEaek:b-yiure4) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/2b396e73357953070924915e7a12a5d8a0e80100/svc1logr/svc1logr.go#L31)):*
> ```Go
> 
> // New returns a go-logr interface implementation that uses svc1log internally.
> func New(logger svc1log.Logger, origin string) logr.Logger {
> ```

Not sure if including `origin string` here is preferable from an API standpoint, especially given that `svc1log` itself supports setting the origin in ways besides using a hard-coded string. Is the motivation to do this to support the `WithName` function? 
___
*[svc1logr/svc1logr.go, line 51 at r3](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MS9r_w1Bu6IP7Jm0Wtm:-MS9r_w1Bu6IP7Jm0Wtn:b-63y5qy) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/2b396e73357953070924915e7a12a5d8a0e80100/svc1logr/svc1logr.go#L51)):*
> ```Go
> }
> 
> func (s *svc1logr) V(level int) logr.InfoLogger {
> ```

Believe the returned interface should be `logr.Logger`.
___
*[svc1logr/svc1logr.go, line 52 at r3](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MS9ktf578WNNWj3o9DV:-MS9ktf578WNNWj3o9DW:b-9xrem) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/2b396e73357953070924915e7a12a5d8a0e80100/svc1logr/svc1logr.go#L52)):*
> ```Go
> 
> func (s *svc1logr) V(level int) logr.InfoLogger {
> 	return New(s.logger, s.origin)
> ```

This doesn't seem to honor the intent of the API -- the intent would be to return a new logger where the service logger level is set based on the provided value. Even if it's not possible to do with the service logger given our API, it should be possible to do this by tracking the level in the `svc1logr` struct and gating output on that.
___
*[svc1logr/svc1logr.go, line 61 at r3](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MSA0YSv1lS2gi0QD9B8:-MSA0YSv1lS2gi0QD9B9:b-vlkcev) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/2b396e73357953070924915e7a12a5d8a0e80100/svc1logr/svc1logr.go#L61)):*
> ```Go
> 
> func (s *svc1logr) WithName(name string) logr.Logger {
> 	return New(s.logger, filepath.Join(s.origin, name))
> ```

This should probably be `path.Join` rather than `filepath.Join`, as I don't think the separator used for the logger should be platform-dependent (don't think that using `\` as the separator on Windows makes sense)


<!-- Sent from Reviewable.io -->

Commit:
logr.Logger

Commit:
verbosity, error log changes, filepath->path

Commit:
better readme

Commit:
move svc1logr -> adapters/svc1logr

etherandrius reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114)* status: 9 of 13 files reviewed, 14 unresolved discussions (waiting on @etherandrius, @nmiyake, and @splucs)

___
*[README.md, line 13 at r3](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MS9j_Am5vfL0UVeoZFV:-MSCrdx40MhzVEgGEUtp:b-896fix) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/2b396e73357953070924915e7a12a5d8a0e80100/README.md#L13)):*
<details><summary><i>Previously, nmiyake (Nick Miyake) wrote…</i></summary><blockquote>

If we proceed with this PR in this repository, then this should be expanded a bit -- I think it would be worth explaining that the repository also provides adapters that will return a particular logger implementation using svc1log underneath, and that an implementation for logr is provided.
</blockquote></details>

Done.
___
*[gologr-wlog/gologr_test.go, line 71 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MQvZ_Sp27m0ErnCysEG-r1-71:-MSCraQT1peOnQNON0Q-:b-896fix) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/c6be7af3baf0bbcd2b983273fdde7eeff1d0e0b7/gologr-wlog/gologr_test.go#L71)):*
<details><summary><i>Previously, splucs (Lucas França de Oliveira) wrote…</i></summary><blockquote>

Remove debug statements.
</blockquote></details>

Done.
___
*[gologr-wlog/gologr.go, line 11 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MQvZ_Sp27m0ErnCysEF-r1-11:-MSCn47_5Xshfn06J1_P:b8jjcni) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/c6be7af3baf0bbcd2b983273fdde7eeff1d0e0b7/gologr-wlog/gologr.go#L11)):*
<details><summary><i>Previously, splucs (Lucas França de Oliveira) wrote…</i></summary><blockquote>

Just name this `logger`.
</blockquote></details>

I prefer to keep this svc1logr. It follows logr convention and doesn't clash with the logger sub-value. 
___
*[gologr-wlog/gologr.go, line 17 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MQvZ_Sp27m0ErnCysEF-r1-17:-MSCn2brC9GxfXClMq70:b-896fix) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/c6be7af3baf0bbcd2b983273fdde7eeff1d0e0b7/gologr-wlog/gologr.go#L17)):*
<details><summary><i>Previously, splucs (Lucas França de Oliveira) wrote…</i></summary><blockquote>

Package name already tells what it is, just name this `New`.
</blockquote></details>

Done.
___
*[gologr-wlog/gologr.go, line 25 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MQvZ_Sp27m0ErnCysEF-r1-25:-MSCmtMw0nBoPT_X2bt_:b-896fix) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/c6be7af3baf0bbcd2b983273fdde7eeff1d0e0b7/gologr-wlog/gologr.go#L25)):*
<details><summary><i>Previously, splucs (Lucas França de Oliveira) wrote…</i></summary><blockquote>

Name the receiver the initial letter of the struct.
</blockquote></details>

Done.
___
*[gologr-wlog/gologr.go, line 52 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MQvZ_Sp27m0ErnCysEF-r1-52:-MSCmTmYCRFk2YLlGmwE:b-qxwonq) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/c6be7af3baf0bbcd2b983273fdde7eeff1d0e0b7/gologr-wlog/gologr.go#L52)):*
<details><summary><i>Previously, nmiyake (Nick Miyake) wrote…</i></summary><blockquote>

When this condition is hit, is it clear that the error is in the logging layer? Agree that it's good to message this in some way, but want to be sure that the source of the error is clear.
</blockquote></details>

- I think origin is enough to signify that this is from the external library. 
- I've prepended the logs with `Logging layer: `
___
*[gologr-wlog/gologr.go, line 57 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MQvZ_Sp27m0ErnCysEF-r1-57:-MSCmrWi32sxSypDnEos:b-896fix) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/c6be7af3baf0bbcd2b983273fdde7eeff1d0e0b7/gologr-wlog/gologr.go#L57)):*
<details><summary><i>Previously, splucs (Lucas França de Oliveira) wrote…</i></summary><blockquote>

`make(map[string]interface{}, len(keysAndValues)/2)`
</blockquote></details>

Done.
___
*[svc1logr/svc1logr.go, line 31 at r3](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MSA19kR6HuWqGXUEaej:-MSCU-C98Ho0j2QvehyY:be6tckk) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/2b396e73357953070924915e7a12a5d8a0e80100/svc1logr/svc1logr.go#L31)):*
<details><summary><i>Previously, nmiyake (Nick Miyake) wrote…</i></summary><blockquote>

Not sure if including `origin string` here is preferable from an API standpoint, especially given that `svc1log` itself supports setting the origin in ways besides using a hard-coded string. Is the motivation to do this to support the `WithName` function? 
</blockquote></details>

Correct `WithName` only appends string to the name, hence we need to store/extract the origin value. I did not find a nice way to extract and change the origin already present in `svc1log.Logger`.
___
*[svc1logr/svc1logr.go, line 51 at r3](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MS9r_w1Bu6IP7Jm0Wtm:-MSCmse4CBC8G-6WEnca:b-896fix) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/2b396e73357953070924915e7a12a5d8a0e80100/svc1logr/svc1logr.go#L51)):*
<details><summary><i>Previously, nmiyake (Nick Miyake) wrote…</i></summary><blockquote>

Believe the returned interface should be `logr.Logger`.
</blockquote></details>

Done.
___
*[svc1logr/svc1logr.go, line 52 at r3](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MS9ktf578WNNWj3o9DV:-MSClwJA8LzfVT0rMZGd:b-7hx010) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/2b396e73357953070924915e7a12a5d8a0e80100/svc1logr/svc1logr.go#L52)):*
<details><summary><i>Previously, nmiyake (Nick Miyake) wrote…</i></summary><blockquote>

This doesn't seem to honor the intent of the API -- the intent would be to return a new logger where the service logger level is set based on the provided value. Even if it's not possible to do with the service logger given our API, it should be possible to do this by tracking the level in the `svc1logr` struct and gating output on that.
</blockquote></details>

Done. 

I've added two values to the struct `level int` and `enabled bool`. 

`level` - signifies up to what verbosity we should log the output out
`enabled` - signifies if the current logging level is below or above our preferred verbosity level.

- added a check to Info(..) function to check if logger is enabled, omitted the same from Error() function since errors are by definition verbosity 0. (It's illegal to have verbosity <0, If a user want to have no logs they should not use this package.)
___
*[svc1logr/svc1logr.go, line 61 at r3](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MSA0YSv1lS2gi0QD9B8:-MSCmS60DPlUX_aSpY6s:b-896fix) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/2b396e73357953070924915e7a12a5d8a0e80100/svc1logr/svc1logr.go#L61)):*
<details><summary><i>Previously, nmiyake (Nick Miyake) wrote…</i></summary><blockquote>

This should probably be `path.Join` rather than `filepath.Join`, as I don't think the separator used for the logger should be platform-dependent (don't think that using `\` as the separator on Windows makes sense)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
fix testss

bmoylan reviewed (commented):
// sorry, meant to publish this yesterday

Thanks @nmiyake for doing a thorough review here.

I actually like the idea of `adapters` -- I've written a few throwaway/internal versions of these over the years and it would be nice to put them in one place. 

One downside is if it bloats the dependency tree. We strive to do a good job keeping the wlog impls isolated, and I'd want to do the same thing here either with modules or just copying the interface and duck-typing it. My gut feeling is that it feels a little heavy and specific to go in `palantir/pkg` and would prefer to keep them in this repo if manageable.

> Reviewed 1 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-:-MSA5T1kBXMzyWeRPMHl:b-rx9zpi)* status: 9 of 13 files reviewed, 15 unresolved discussions (waiting on @bmoylan, @etherandrius, @nmiyake, and @splucs)

___
*[svc1logr/svc1logr.go, line 30 at r3](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MSA6dZHA8qKC-0E3LQs:-MSA6dZHA8qKC-0E3LQt:bcarzfj) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/2b396e73357953070924915e7a12a5d8a0e80100/svc1logr/svc1logr.go#L30)):*
> ```Go
> }
> 
> // New returns a go-logr interface implementation that uses svc1log internally.
> ```

Maybe a better doc is something like
```
// New provides an implementation of [logr.Logger](https://pkg.go.dev/github.com/go-logr/logr#Logger) which delegates to a witchcraft svc1log logger.
// Use this package for software which expects a logr.Logger but where the calling code has access to an svc1log.Logger or wants the output to be Witchcraft-compatible.
// V levels are ignored: Info() is logged as INFO and Error() as ERROR. All values are converted to safe parameters.
```


<!-- Sent from Reviewable.io -->

Commit:
better package docs

etherandrius reviewed (commented):
I think I've addressed all the changes for now. 

The only thing that's up in the air is where this should live:
- In a top level package `adapters`
- OR; In a svc1log subpackage `svclog/svc1log/adapters/svc1logr`

For now I am keeping it as a top level package. 

As for duck-typing: I do not love the idea. For me it seems that it will only bring confusion and obfuscate the code. But It's up to you guys to decide I won't fight on the decision much. 

> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-:-MSSJzAM3mJpbbwKZd3z:b-77vdli)* status: 9 of 12 files reviewed, 15 unresolved discussions (waiting on @bmoylan, @etherandrius, @nmiyake, and @splucs)

___
*[svc1logr/svc1logr.go, line 30 at r3](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MSA6dZHA8qKC-0E3LQs:-MSSJmQi6PMK-wF0t_Y9:b-5pxwur) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/2b396e73357953070924915e7a12a5d8a0e80100/svc1logr/svc1logr.go#L30)):*
<details><summary><i>Previously, bmoylan (Brad Moylan) wrote…</i></summary><blockquote>

Maybe a better doc is something like
```
// New provides an implementation of [logr.Logger](https://pkg.go.dev/github.com/go-logr/logr#Logger) which delegates to a witchcraft svc1log logger.
// Use this package for software which expects a logr.Logger but where the calling code has access to an svc1log.Logger or wants the output to be Witchcraft-compatible.
// V levels are ignored: Info() is logged as INFO and Error() as ERROR. All values are converted to safe parameters.
```
</blockquote></details>

Done. Modified the comment  slightly since we now support V levels, to be more inline with the API. 


<!-- Sent from Reviewable.io -->

nmiyake reviewed (changes requested):
Sorry for the long delay.

Reviewed again, and am okay with the overall concept/approach. I would prefer the current location in the top-level "adapters" directory over mingling individual adapters as subpackages.

Left feedback on content.

> Reviewed 3 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-:-MUQBuHY4pSq6_Ihhy2j:b-qzls25)* status: all files reviewed, 15 unresolved discussions (waiting on @bmoylan, @etherandrius, @nmiyake, and @splucs)

___
*[README.md, line 9 at r4](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MUQBmnw3u8cwsrnKUGh:-MUQBmnx4vTOyfumn1vc:b4pimn1) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/75996bb3967e97d57157ae4691f0d79ca77b2c17/README.md#L9)):*
> ```Markdown
> [![](https://godoc.org/github.com/palantir/witchcraft-go-logging?status.svg)](http://godoc.org/github.com/palantir/witchcraft-go-logging)
> 
> `witchcraft-go-logging` is a Go implementationof the Witchcraft logging specification. It provides an API that can be
> ```

Would prefer this for this section:

```
`witchcraft-go-logging` is a Go implementation of the Witchcraft logging specification. It provides an API that can be
used for logging and some default implementations of the logging API using different existing popular Go logging
libraries.

Currently, implementations that use the following libraries are provided:
* [zap](https://github.com/uber-go/zap)
* [zerolog](https://github.com/rs/zerolog)
* [glog](https://github.com/golang/glog)

This module also includes adapters that allow Witchcraft loggers to satisfy a different logging interface. The most
common use case is provide compatibility between the `svc1log.Logger` interface and other Go logging interfaces.

Currently, adapters for the following interfaces are provided:
* [go-logr](https://github.com/go-logr/logr)
```
___
*[adapters/svc1logr/svc1logr.go, line 42 at r4](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MUQ5nyG7sr6mNnZJrsU:-MUQ5nyG7sr6mNnZJrsV:bb7wk8f) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/75996bb3967e97d57157ae4691f0d79ca77b2c17/adapters/svc1logr/svc1logr.go#L42)):*
> ```Go
> 		logger:  logger,
> 		level:   level,
> 		enabled: true,
> ```

This results in incorrect behavior if the provided level is  >0, since the logger will behave as if it is enabled.
___
*[adapters/svc1logr/svc1logr.go, line 60 at r4](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MUQ9LDy5wkaDrftzanW:-MUQ9LDy5wkaDrftzanX:bk9938j) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/75996bb3967e97d57157ae4691f0d79ca77b2c17/adapters/svc1logr/svc1logr.go#L60)):*
> ```Go
> }
> 
> func (s *svc1logr) V(level int) logr.Logger {
> ```

Would add comment to both constructor and to this function that explains that the level being set here only controls whether or not this interface will perform the logging, and is independent from the underlying logger (if the `svc1log.Logger` being wrapped is at the ERROR level, then it won't output anything at INFO level regardless of the level set here)
___
*[adapters/svc1logr/svc1logr.go, line 68 at r4](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MUQ8kVx3-UPz08HS5cI:-MUQ8kVx3-UPz08HS5cJ:b-v5144t) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/75996bb3967e97d57157ae4691f0d79ca77b2c17/adapters/svc1logr/svc1logr.go#L68)):*
> ```Go
> 		origin:  s.origin,
> 		logger:  s.logger,
> 		level:   s.level,
> ```

This implementation still doesn't match the interface API -- based on the documentation and examples of existing implementations, I think this is meant to be strictly additive:

* https://github.com/go-logr/glogr/blob/master/glogr.go#L198
* https://github.com/go-logr/zapr/blob/master/zapr.go#L139
___
*[svc1logr/svc1logr.go, line 52 at r3](https://reviewable.io/reviews/palantir/witchcraft-go-logging/114#-MS9ktf578WNNWj3o9DV:-MUQ8RMb7Z4Z9fEeLFKq:b-jwaklh) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/2b396e73357953070924915e7a12a5d8a0e80100/svc1logr/svc1logr.go#L52)):*
<details><summary><i>Previously, etherandrius (Andrius Grabauskas) wrote…</i></summary><blockquote>

Done. 

I've added two values to the struct `level int` and `enabled bool`. 

`level` - signifies up to what verbosity we should log the output out
`enabled` - signifies if the current logging level is below or above our preferred verbosity level.

- added a check to Info(..) function to check if logger is enabled, omitted the same from Error() function since errors are by definition verbosity 0. (It's illegal to have verbosity <0, If a user want to have no logs they should not use this package.)
</blockquote></details>

Thanks for this information -- would add this as a comment in the implementation of the `Error` function so that this context is captured in code.


<!-- Sent from Reviewable.io -->
